### DISCUSSION TITLE
Troubles getting fpga code to update on device

### DISCUSSION THREAD

**Discussion by d3x0r** [HerrNamenlos123/JTAG_Interface#8] [General]
This is my forked version, https://github.com/d3x0r/JTAG_Interface/tree/d3x0r-hispeed-counter  (on branch d3x0r-hispeed-counter)

https://github.com/d3x0r/JTAG_Interface/tree/d3x0r-hispeed-counter/src  This is where FPGA_Bitstream.h is - I had renamed the old one to .original, and copied the reversed MKRVIDOR4000.ttf  (reversed using the ONLY_CLI version of the reverser which was built by hand), into the FPGA_Bitstream.h.  Having downloaded the released version that runs in the systray, the bitstream.h file is the same as what I had replaced - so there was nothing wrong there...

I do get
```
JTAG initialized
A0 = 1
A1 = 1
A2 = 1

The FPGA program has been running for 0 115842603 0 ticks, that is 0.00 seconds.
```
as the first logging in the Arduino IDE after trying to reupload the file...

(The logging has changed, so the first 0 should be the 1158... value, and the others should be 0 until I send a command)  So I know the .ino is updated  https://github.com/d3x0r/JTAG_Interface/blob/d3x0r-hispeed-counter/examples/simple/simple.ino#L64  and gets downloaded to the board, but I did move the old counter in Quartus from 9 to 7, and inserted a connection to my new counter at 9...

![image](https://github.com/HerrNamenlos123/JTAG_Interface/assets/1909610/dec16145-60d0-4de0-989f-279ac19951b1)

Hence, the 'ticks' variable in the Arduino project should be pulling the old timer, but the FPGA code is still generating the data on 9 where it was originally wired.... 

There was a thread somewhere I ran into that said something like 'had to reflash to original loader' and then re-flash with their new program to get the FPGA to update... But I'm not sure where or really even if that applied to this board

**Comment by d3x0r** [comment]
Okay - I do recall I had to install a library in arduino - it was failing to find 'FPGA.h' - which is in /src/ in your project... but if I rename that file, then it still compiles in the arduino IDE... so I guess there's a different for fpga.h  like ` /users/me/documents/Arduino/libraries/JTAG_Interface/src`  Which then if I replace the FPGA_Bitstream.h in THAT folder I get a different error entirely.

JTAG FPGA mismatch. Error:
The number of usable registers does not match the JTAG module on the FPGA. Make sure the right FPGA bitstream is being loaded.

**Reply by HerrNamenlos123** [reply]
Ah, this error is to be expected. It means that the Arduino has a different knowledge of JTAG than the FPGA. The FPGA JTAG instance is configurable and you can change it. However, the Arduino needs to know that, so it can correctly check out of bounds writes. In your case that means that you probably changed the FPGA code (or didn't) and the Arduino code has wrong numbers in FPGA.begin(). You must change the numbers in FPGA.begin() so that they match the FPGA instance.
